### PR TITLE
Add basic auth support to api.py

### PR DESCRIPTION
**Pull request by pico-prob**
Hi Bernhard, I use your Signal CLI service container (tnx, btw!) and put it behind a nginx with an basic auth protection. Now I adapted your python client API. It's a rather superficial adaption (just manual tests) - maybe you'll have a look at it yourself. And I am not that proficient with python..

One more note: my change always adds a `basic_auth` parameter to `requests` call. Without basic auth parameters it defaults to `None`. This matches the [current `requests` API](https://docs.python-requests.org/en/latest/api/#requests.Request). 

It would be of great help if you push the update to pypi.

Best!
Mathias

### CODE REVIEW

**Comment by bbernhard** [comment]
Thanks for the PR!

Something I would prefer is if we could make the authentication more generic, in order to support all kind of different authentication mechanisms. I guess that there are probably users out there that want to use different authentication mechanisms (like token based authentication). `requests` already supports a bunch of authentication methods and even allows you to create [your own authentication method](https://docs.python-requests.org/en/latest/user/authentication/#new-forms-of-authentication) by subclassing `requests.auth.AuthBase`. I haven't tried it yet, but it should be possible to just pass such an authentication object into the Python constructor and use that one. That way we could support all kind of different authentication methods without cluttering the API.

**Comment by pico-prob** [comment]
Hi Bernhard, if I get you right this would just change the constructor parameters: Instead of two Strings (`basic_auth_user, basic_auth_pwd`) there would be one `requests.auth.AuthBase` parameter.

For my purposes this would be fine - actually better as more flexible in the future. However it would bring another dependency to the API client. This could be avoided, if we keep both (i.e. 3 constructor parameter, each with default). It would clutter the code though.. I can't say what's better. What do you think?

**Comment by bbernhard** [comment]
What I am a bit afraid of is, that we end up with a constructor that looks like this:

`def __init__(self, base_url, number, basic_auth_user, basic_auth_pwd. jwt_token, oauth, ...):`

Even if all those arguments are defaulted to `None`, we end up with a constructor that's pretty cluttered. 

As we already require `requests` to be installed, I think we could use their authentication classes to make our lives easier.

I am imagening something like this:

```
from pysignalclirestapi import SignalCliRestApi
from requests.auth import HTTPBasicAuth

auth = HTTPBasicAuth('user', 'pass')
api = SignalCliRestApi("http://127.0.0.1:8080", "+123456789", auth=auth)
api.send("+4123456779")
```

If one wants to change the authentication method, all that needs to be done is to use a different authentication class:

```
from pysignalclirestapi import SignalCliRestApi
from requests.auth import HTTPDigestAuth

auth = HTTPDigestAuth('user', 'pass')
api = SignalCliRestApi("http://127.0.0.1:8080", "+123456789", auth=auth)
api.send("+4123456779")
```

If we want to hide the fact that we are using `requests` internally, I guess we could create our own wrapper (e.g: `SignalCliRestApiBasicAuth`, `SignalCliRestApiOAuth1Auth`, etc.) classes which inherit from the appropriate `requests` classes.

**Comment by pico-prob** [comment]
Hi Bernhard, I liked your idea with the wrapper. I created an abstract class `SignalCliRestApiAuth` that the client may subclass for their own (requests) authentication. The api also includes an wrapper for basic-auth. It might look a bit java style.. 😬
Maybe we should put the auth classes into their own file?

File: pysignalclirestapi/__init__.py [hunk 1]
```diff
@@ -1 +1 @@
-from pysignalclirestapi.api import SignalCliRestApi, SignalCliRestApiError
+from pysignalclirestapi.api import SignalCliRestApi, SignalCliRestApiError, SignalCliRestApiAuth, SignalCliRestApiAuth, SignalCliRestApiHTTPBasicAuth
```
Comment (bbernhard) [inline pysignalclirestapi/__init__.py:1]: I think you have the `SignalCliRestApiAuth` twice here?

File: pysignalclirestapi/api.py [hunk 2]
```diff
@@ -3,35 +3,55 @@
...
-    """SignalCliRestApiError base classi."""
+    """SignalCliRestApiError base class."""
     pass
 
 
+class SignalCliRestApiAuth(ABC):
+    """SignalCliRestApiAuth base class."""
+
+    @abstractmethod
+    def getAuth():
```
Comment (bbernhard) [inline pysignalclirestapi/api.py:22]: Currently all the methods are in snake case. So I think we should rename that to `get_auth` here as well.

File: pysignalclirestapi/api.py [hunk 3]
```diff
@@ -3,35 +3,55 @@
...
+        pass
+
+
+class SignalCliRestApiHTTPBasicAuth(SignalCliRestApiAuth):
+    """SignalCliRestApiHTTPBasicAuth offers HTTP basic authentication."""
+
+    def __init__(self, basic_auth_user, basic_auth_pwd):
+        self._auth = HTTPBasicAuth(basic_auth_user, basic_auth_pwd)
+
+    def getAuth(self):
```
Comment (bbernhard) [inline pysignalclirestapi/api.py:32]: same here.

**Review by bbernhard — COMMENTED** [review]
Sorry for the late reply. 

I found a few small cosmetic improvements, but apart from that I think it looks good!

Many thanks!

**Comment by pico-prob** [comment]
Agreed & done!

**Comment by bbernhard** [comment]
Looks good, thanks a lot!

**Comment by pico-prob** [comment]
Thank you for the container + client!

**Comment by robin-thoni** [comment]
Quite "old" and merged already, but still:
- A pattern I saw already is to allow the user to pass a `Session` object. That way, the user is free to customize and share the session w/ other code/libraries. If the user does not pass anything (`None`), then simply init a new `Session`. This would also remove the `verify_ssl` argument.
- What's the point of `SignalCliRestApiAuth`? It does not provide any feature. It could simply be replaced by a functor. Even then, I still don't see any point, as it will be executed right after, w/o any callback argument. So the result is predictable anyway.

**Comment by pico-prob** [comment]
Hey @robin-thoni, I'm not quite sure if I fully understand your concern:
You want to use the [`requests.Session`](https://docs.python-requests.org/en/master/user/advanced/#session-objects) type object in place of the [`requests.auth.*`](https://docs.python-requests.org/en/master/user/authentication/#authentication) as it contains both authentication and call (ie `get`, `post`,..)?

The motive ("feature") of `SignalCliRestApiAuth` is to not create a dependency to the `requests` module for someone not caring for authentication. I think callbacks or session sharing has just been out of scope so far...

**Comment by robin-thoni** [comment]
> You want to use the [requests.Session](https://docs.python-requests.org/en/master/user/advanced/#session-objects) type object in place of the [requests.auth.*](https://docs.python-requests.org/en/master/user/authentication/#authentication) as it contains both authentication and call (ie get, post,..)?

Correct.

> The motive ("feature") of SignalCliRestApiAuth is to not create a dependency to the requests module for someone not caring for authentication.

Sorry, I've been unclear here: I do understand the _goal_, but I do not understand _why_ such a goal: that's python :) The dependency will be installed, that's for sure, so I don't understand why you want to hide it. Instead of passing a `SignalCliRestApiAuth` to `__init__`, you'd just pass an optional `Session` object. This gives much more flexibility.

> I think callbacks or session sharing has just been out of scope so far...

To be clear, what I meant by callback is that an interface with a single method can be replaced by a simple "function pointer"/`typing.Callable`

**Pull request merged**